arangod/Aql/SortedRowsStorageBackendMemory.cpp

```diff
@@ -0,0 +1,218 @@
+    // This executor is passthrough. it has enough place to write.
+    _rowIndexes.emplace_back(
+        std::make_pair(static_cast<std::uint32_t>(_inputBlocks.size() - 1),
+                       static_cast<std::uint32_t>(inputRange.getRowIndex())));
+    std::tie(state, input) =
+        inputRange.nextDataRow(AqlItemBlockInputRange::HasDataRow{});
+    TRI_ASSERT(input.isInitialized());
+  }
+
+  guard.steal();
+
+  return state;
+}
+
+bool SortedRowsStorageBackendMemory::hasReachedCapacityLimit() const noexcept {
+  return (_rowIndexes.size() > _infos.spillOverThresholdNumRows() ||
+          _memoryUsageForInputBlocks > _infos.spillOverThresholdMemoryUsage());
+}
+
+bool SortedRowsStorageBackendMemory::hasMore() const {
+  TRI_ASSERT(_sealed);
+  return _returnNext < _rowIndexes.size();
+}
+
+void SortedRowsStorageBackendMemory::produceOutputRow(
+    OutputAqlItemRow& output) {
+  TRI_ASSERT(hasMore());
+  InputAqlItemRow inRow(_inputBlocks[_rowIndexes[_returnNext].first],
+                        _rowIndexes[_returnNext].second);
+  output.copyRow(inRow);
+  output.advanceRow();
+  ++_returnNext;
+}
+
+void SortedRowsStorageBackendMemory::skipOutputRow() noexcept {
+  TRI_ASSERT(hasMore());
+  ++_returnNext;
+}
+
+void SortedRowsStorageBackendMemory::seal() {
+  TRI_ASSERT(!_sealed);
+  doSorting();
+  _sealed = true;
+}
+
+void SortedRowsStorageBackendMemory::spillOver(
+    SortedRowsStorageBackend& other) {
+  if (_rowIndexes.empty()) {
+    return;
+  }
+
+  std::uint32_t lastBlockId = _rowIndexes[0].first;
+  std::uint32_t startRow = _rowIndexes[0].second;
+
+  for (size_t i = 0; i < _rowIndexes.size(); ++i) {
+    if (_rowIndexes[i].first != lastBlockId || i == _rowIndexes.size() - 1) {
+      // emit block from lastBlockId to startRow, endRow
+      aql::AqlItemBlockInputRange inputRange(
+          aql::MainQueryState::HASMORE, 0, _inputBlocks[lastBlockId], startRow);
+
+      other.consumeInputRange(inputRange);
+
+      lastBlockId = _rowIndexes[i].first;
+      startRow = _rowIndexes[i].second;
+    }
+  }
+
+  // reset our own state, so we can give back memory
+  _infos.getResourceMonitor().decreaseMemoryUsage(currentMemoryUsage());
+  _inputBlocks = {};
+  _rowIndexes.clear();
+  _rowIndexes.shrink_to_fit();
+  TRI_ASSERT(currentMemoryUsage() == 0);
+  _memoryUsageForInputBlocks = 0;
+}
+
+void SortedRowsStorageBackendMemory::doSorting() {
+  TRI_IF_FAILURE("SortBlock::doSorting") {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+  }
+
+  // comparison function
+  OurLessThan ourLessThan(_infos.vpackOptions(), _inputBlocks,
+                          _infos.sortRegisters());
+  if (_infos.stable()) {
+    std::stable_sort(_rowIndexes.begin(), _rowIndexes.end(), ourLessThan);
+  } else {
+    std::sort(_rowIndexes.begin(), _rowIndexes.end(), ourLessThan);
+  }
+}
+
+size_t SortedRowsStorageBackendMemory::currentMemoryUsage() const noexcept {
+  return _rowIndexes.capacity() * sizeof(AqlItemMatrix::RowIndex);
```

Comment:
I would say not because the `_memoryUsageForInputBlocks` is used as a value to compare against the memory usage threshold that we can provide for opting to use rocksdb for temporary storage, so it's only used for this comparison not as contribution to the memory. So every time there's more input, we increase this and check if shouldn't use rocksdb yet

What is used as contributing to the resource monitor memory usage is the vector of row indexes

Reply:
OK, understood, so this is OK.